File: impl/java/src/main/java/com/cptnfizzbin/keycard/lib/Catalog.java
package com.cptnfizzbin.keycard.lib;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import com.cptnfizzbin.keycard.errors.PolicyArgumentException;

/**
 * Resolves a {@code KeycardConfig}'s {@code actions}/{@code subjects}
 * catalog into the reverse {@code raw name -> catalog key} map {@code
 * PolicyBuilder}/{@code Policy} use to resolve a dynamic Action/Subject's
 * random name into its real, serializable one, plus the full list of names
 * to fold into {@code meta.actions}/{@code meta.subjects}.
 */
public final class Catalog {
    private Catalog() {}

    /**
     * @param reverseMap raw name (a dynamic Action/Subject's random id, or a
     *   named one's own name) -> catalog key.
     * @param names every catalog key, in registration order.
     */
    public record Resolution(Map<String, String> reverseMap, List<String> names) {}

    /**
     * @param catalog the keyed catalog (e.g. {@code
     *   KeycardConfig.actions().asMap()}) - each key becomes the serialized
     *   name for its entry.
     * @param nameOf reads an entry's own (possibly dynamic/random) name.
     * @param kind used only to name the vocabulary ("action"/"subject") in
     *   a duplicate-registration error message.
     */
    public static <T> Resolution build(Map<String, ? extends T> catalog, Function<T, String> nameOf, String kind) {
        List<String> names = new ArrayList<>();
        Map<String, String> reverseMap = new LinkedHashMap<>();
        if (catalog != null) {
            for (Map.Entry<String, ? extends T> entry : catalog.entrySet()) {
                String key = entry.getKey();
                String rawName = nameOf.apply(entry.getValue());
                String existingKey = reverseMap.get(rawName);
                if (existingKey != null && !existingKey.equals(key)) {
                    throw new PolicyArgumentException(
                        "KeycardConfig " + kind + " catalog error: the same " + kind + " is registered under both \""
                            + existingKey + "\" and \"" + key
                            + "\" - a single Action/Subject can only be registered under one catalog key."
                    );
                }
                reverseMap.put(rawName, key);
                names.add(key);
            }
        }

        return new Resolution(reverseMap, names);
    }

    /** Resolves {@code rawName} (an Action/Subject's own name, dynamic or not) to its catalog key, or returns it unchanged when it isn't a registered catalog entry. */
    public static String resolveName(Map<String, String> reverseMap, String rawName) {
        String resolved = reverseMap.get(rawName);
        return resolved != null ? resolved : rawName;
    }
}

File: impl/js/src/lib/catalog.ts
import { PolicyArgumentError } from "../errors/index.ts"

/**
 * Resolves a `KeycardConfig.actions`/`.subjects` catalog (an `ActionCatalog`/
 * `SubjectCatalog`, i.e. a keyed `Record<string, T>`) into the reverse
 * `id -> catalog key` map `PolicyBuilder`/`Policy` use to resolve a dynamic
 * Action/Subject's random name into its real, serializable one, plus the
 * full list of names to fold into `meta.actions`/`meta.subjects`.
 */
export interface CatalogResolution {
  /** Raw name (a dynamic Action/Subject's random id, or a named entry's own name) -> catalog key. */
  reverseMap: Map<string, string>
  /** Every resolved name: each entry's catalog key. */
  names: string[]
}

const EMPTY_RESOLUTION: CatalogResolution = { reverseMap: new Map(), names: [] }

/**
 * @param kind used only to name the vocabulary ("action"/"subject") in a
 *   duplicate-registration error message.
 * @param validate when false (`KeycardConfig.emitMeta: false`), skips the
 *   duplicate-key check below - a definition that already passed CI once
 *   doesn't need to re-prove itself on every construction. The last entry
 *   for a given raw name still wins in `reverseMap`, silently.
 */
export function buildCatalog<T extends { name: string }>(
  entries: Record<string, T> | undefined,
  kind: string,
  validate = true,
): CatalogResolution {
  if (entries === undefined) return EMPTY_RESOLUTION

  const reverseMap = new Map<string, string>()
  const names: string[] = []
  for (const [key, entry] of Object.entries(entries)) {
    if (validate) {
      const existingKey = reverseMap.get(entry.name)
      if (existingKey !== undefined && existingKey !== key) {
        throw new PolicyArgumentError(
          `KeycardConfig ${kind} catalog error: the same ${kind} is registered under both "${existingKey}" and "${key}" - a single Action/Subject can only be registered under one catalog key.`,
        )
      }
    }
    reverseMap.set(entry.name, key)
    names.push(key)
  }

  return { reverseMap, names }
}

/** Resolves `rawName` (an Action/Subject's `.name`, dynamic or not) to its catalog key, or returns it unchanged when it isn't a registered catalog entry. */
export function resolveName(reverseMap: Map<string, string>, rawName: string): string {
  return reverseMap.get(rawName) ?? rawName
}
